From: Gaylon Buckelew <gaylon.buckelew@adtran.com>
Date: Fri, 29 Sep 2023 00:00:00 +0000
Subject: [PATCH] hostapd: expose detailed WiFi client capabilities via ubus

Extend the ubus interface to report comprehensive client capabilities
including HT, VHT, HE, and EHT parameters. Adds get_client_caps method
for capabilities-only queries and vendor IE collection.

Co-developed-by: Vitali Klotschko <vitali.klotschko@adtran.com>
Signed-off-by: Vitali Klotschko <vitali.klotschko@adtran.com>
Co-developed-by: Matthew Cather <matthew.cather@adtran.com>
Signed-off-by: Matthew Cather <matthew.cather@adtran.com>
Signed-off-by: Gaylon Buckelew <gaylon.buckelew@adtran.com>
Signed-off-by: Chad Monroe <chad@monroe.io>
---
 src/ap/ieee802_11.c            |    4 
 src/ap/ieee802_11_eht.c        |    3 
 src/ap/sta_info.c              |    1 
 src/ap/sta_info.h              |    2 
 src/ap/ubus.c                  |  714 +++++++++++++++++++++++++++++--
 src/common/ieee802_11_common.c |   27 +
 src/common/ieee802_11_common.h |    1 
 7 files changed, 710 insertions(+), 42 deletions(-)

--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -6152,6 +6152,14 @@ skip_wpa_ies:
 		hostapd_wfa_capab(hapd, sta, elems->wfa_capab,
 				  elems->wfa_capab + elems->wfa_capab_len);
 
+	wpabuf_free(sta->vendors_ie);
+	sta->vendors_ie = (ies && ies_len) ?
+		ieee802_11_all_vendor_ie_concat(ies, ies_len) : NULL;
+
+	wpabuf_free(sta->assoc_ies);
+	sta->assoc_ies = (ies && ies_len) ?
+		wpabuf_alloc_copy(ies, ies_len) : NULL;
+
 out:
 	if (resp != WLAN_STATUS_SUCCESS || assoc_wpa_sm) {
 		struct wpa_state_machine *sm = sta->wpa_sm;
--- a/src/ap/sta_info.c
+++ b/src/ap/sta_info.c
@@ -480,6 +480,8 @@ void ap_free_sta(struct hostapd_data *ha
 	os_free(sta->t_c_url);
 	wpabuf_free(sta->hs20_deauth_req);
 	os_free(sta->hs20_session_info_url);
+	wpabuf_free(sta->vendors_ie);
+	wpabuf_free(sta->assoc_ies);
 
 #ifdef CONFIG_SAE
 	sae_clear_data(sta->sae);
--- a/src/ap/sta_info.h
+++ b/src/ap/sta_info.h
@@ -350,6 +350,9 @@ struct sta_info {
 	struct link_reconf_req_list *reconf_req;
 #endif /* CONFIG_IEEE80211BE */
 
+	struct wpabuf *vendors_ie;
+	struct wpabuf *assoc_ies;
+
 	u16 max_idle_period; /* if nonzero, the granted BSS max idle period in
 			      * units of 1000 TUs */
 
--- a/src/common/ieee802_11_common.c
+++ b/src/common/ieee802_11_common.c
@@ -1283,6 +1283,33 @@ struct wpabuf * ieee802_11_vendor_ie_con
 }
 
 
+struct wpabuf * ieee802_11_all_vendor_ie_concat(const u8 *ies, size_t ies_len)
+{
+	struct wpabuf *buf;
+	const struct element *elem;
+	size_t total_len = 0;
+
+	for_each_element_id(elem, WLAN_EID_VENDOR_SPECIFIC, ies, ies_len) {
+		if (elem->datalen >= 4)
+			total_len += elem->datalen + 2;
+	}
+
+	if (total_len == 0)
+		return NULL;
+
+	buf = wpabuf_alloc(total_len);
+	if (!buf)
+		return NULL;
+
+	for_each_element_id(elem, WLAN_EID_VENDOR_SPECIFIC, ies, ies_len) {
+		if (elem->datalen >= 4)
+			wpabuf_put_data(buf, &elem->id, elem->datalen + 2);
+	}
+
+	return buf;
+}
+
+
 const u8 * get_hdr_bssid(const struct ieee80211_hdr *hdr, size_t len)
 {
 	u16 fc, type, stype;
--- a/src/common/ieee802_11_common.h
+++ b/src/common/ieee802_11_common.h
@@ -237,6 +237,7 @@ ParseRes ieee802_11_parse_link_assoc_res
 int ieee802_11_ie_count(const u8 *ies, size_t ies_len);
 struct wpabuf * ieee802_11_vendor_ie_concat(const u8 *ies, size_t ies_len,
 					    u32 oui_type);
+struct wpabuf * ieee802_11_all_vendor_ie_concat(const u8 *ies, size_t ies_len);
 struct ieee80211_hdr;
 const u8 * get_hdr_bssid(const struct ieee80211_hdr *hdr, size_t len);
 
--- a/src/ap/ubus.c
+++ b/src/ap/ubus.c
@@ -11,6 +11,7 @@
 #include "utils/eloop.h"
 #include "utils/wpabuf.h"
 #include "common/ieee802_11_defs.h"
+#include "common/ieee802_11_common.h"
 #include "common/hw_features_common.h"
 #include "hostapd.h"
 #include "neighbor_db.h"
@@ -25,6 +26,19 @@
 #include "airtime_policy.h"
 #include "hw_features.h"
 #include "base64.h"
+#include <stdlib.h>
+
+#define MCS_NSS_MAP_BASE		7
+#define MCS_NSS_NOT_SUPPORTED		(-1)
+#define HT_MCS_32			32
+#define HT_MAX_SPATIAL_STREAMS		4
+#define HT_TX_MCS_SET_DEFINED_BIT	96
+#define HT_TX_RX_MCS_NOT_EQUAL_BIT	97
+#define HT_TX_MAX_NSS_BIT		98
+#define HT_TX_UEQM_BIT			100
+#define VHT_CAP_EXTENDED_NSS_BW_SHIFT	30
+
+u16 smartrg_eht_ppet_size(u16 ppe_thres_hdr, const u8 *phy_cap_info);
 
 static struct ubus_context *ctx;
 static struct blob_buf b;
@@ -169,77 +183,689 @@ hostapd_bss_reload(struct ubus_context *
 
 
 static void
-hostapd_parse_vht_map_blobmsg(uint16_t map)
+mcs_nss_map_add(uint16_t map, u32 flags)
 {
-	char label[4];
-	int16_t val;
-	int i;
+	bool is_vht = (flags & WLAN_STA_VHT) != 0;
+	int mult = is_vht ? 1 : 2;
+	int16_t unsupported_val = 3 * mult + MCS_NSS_MAP_BASE;
 
-	for (i = 0; i < 8; i++) {
-		snprintf(label, 4, "%dss", i + 1);
+	for (int i = 0; i < VHT_RX_NSS_MAX_STREAMS; i++, map >>= 2) {
+		char label[4];
+		int16_t val = (map & 0x3) * mult + MCS_NSS_MAP_BASE;
 
-		val = (map & (BIT(1) | BIT(0))) + 7;
-		blobmsg_add_u16(&b, label, val == 10 ? -1 : val);
-		map = map >> 2;
+		snprintf(label, sizeof(label), "%dss", i + 1);
+		blobmsg_add_u16(&b, label, val == unsupported_val ? MCS_NSS_NOT_SUPPORTED : val);
 	}
 }
 
+typedef struct bitfield_descriptor {
+	const char *name;
+	uint8_t bitstart;
+	uint8_t bitcount;
+} bitfield_descriptor;
+
+typedef struct {
+	const char *name;
+	uint8_t min_mcs;
+	uint8_t max_mcs;
+} ht_mcs_t;
+
+
 static void
-hostapd_parse_vht_capab_blobmsg(struct ieee80211_vht_capabilities *vhtc)
+vht_capab_add(struct ieee80211_vht_capabilities *vhtc, bool filter_false)
 {
-	void *supported_mcs;
-	void *map;
-	int i;
-
 	static const struct {
 		const char *name;
 		uint32_t flag;
-	} vht_capas[] = {
+	} flags[] = {
+		{ "rx_ldpc", VHT_CAP_RXLDPC },
+		{ "short_gi_80MHz_tvht_mode_4c", VHT_CAP_SHORT_GI_80 },
+		{ "short_gi_160_80_80MHz", VHT_CAP_SHORT_GI_160 },
+		{ "tx_stbc", VHT_CAP_TXSTBC },
+		{ "su_beamformer", VHT_CAP_SU_BEAMFORMER_CAPABLE },
 		{ "su_beamformee", VHT_CAP_SU_BEAMFORMEE_CAPABLE },
+		{ "mu_beamformer", VHT_CAP_MU_BEAMFORMER_CAPABLE },
 		{ "mu_beamformee", VHT_CAP_MU_BEAMFORMEE_CAPABLE },
+		{ "tx_op_ps", VHT_CAP_VHT_TXOP_PS },
+		{ "htc_vht", VHT_CAP_HTC_VHT },
+		{ "vht_link_adaptation_mrq_mfb", VHT_CAP_VHT_LINK_ADAPTATION_VHT_MRQ_MFB },
+		{ "vht_link_adaptation_unsol_mfg", VHT_CAP_VHT_LINK_ADAPTATION_VHT_UNSOL_MFB },
+		{ "rx_antenna_pattern", VHT_CAP_RX_ANTENNA_PATTERN },
+		{ "tx_antenna_pattern", VHT_CAP_TX_ANTENNA_PATTERN }
 	};
 
-	for (i = 0; i < ARRAY_SIZE(vht_capas); i++)
-		blobmsg_add_u8(&b, vht_capas[i].name,
-				!!(vhtc->vht_capabilities_info & vht_capas[i].flag));
+	static const struct {
+		const char *name;
+		uint32_t mask;
+		uint32_t shift;
+	} values[] = {
+		{ "max_mpdu_length", VHT_CAP_MAX_MPDU_LENGTH_MASK, 0 },
+		{ "channel_width_set", VHT_CAP_SUPP_CHAN_WIDTH_MASK, VHT_CAP_SUPP_CHAN_WIDTH_MASK_SHIFT },
+		{ "rx_stbc", VHT_CAP_RXSTBC_MASK, VHT_CAP_RXSTBC_MASK_SHIFT },
+		{ "beamformee_sts_max", VHT_CAP_BEAMFORMEE_STS_MAX, VHT_CAP_BEAMFORMEE_STS_MAX_SHIFT },
+		{ "sounding_dimension_max", VHT_CAP_SOUNDING_DIMENSION_MAX, VHT_CAP_SOUNDING_DIMENSION_MAX_SHIFT },
+		{ "max_ampdu_length_exponent", VHT_CAP_MAX_A_MPDU_LENGTH_EXPONENT_MAX, VHT_CAP_MAX_A_MPDU_LENGTH_EXPONENT_MAX_SHIFT },
+		{ "extended_nss_bw", VHT_CAP_EXTENDED_NSS_BW_SUPPORT_MASK, VHT_CAP_EXTENDED_NSS_BW_SHIFT },
+	};
+
+	uint32_t cap_info = vhtc->vht_capabilities_info;
+	void *mcs_table, *map;
+
+	for (int i = 0; i < ARRAY_SIZE(flags); i++) {
+		uint8_t val = !!(cap_info & flags[i].flag);
+		if (filter_false && !val)
+			continue;
+		blobmsg_add_u8(&b, flags[i].name, val);
+	}
+
+	for (int i = 0; i < ARRAY_SIZE(values); i++) {
+		uint16_t val = (cap_info & values[i].mask) >> values[i].shift;
+		blobmsg_add_u32(&b, values[i].name, val);
+	}
 
-	supported_mcs = blobmsg_open_table(&b, "mcs_map");
+	mcs_table = blobmsg_open_table(&b, "mcs_map");
 
-	/* RX map */
 	map = blobmsg_open_table(&b, "rx");
-	hostapd_parse_vht_map_blobmsg(le_to_host16(vhtc->vht_supported_mcs_set.rx_map));
+	mcs_nss_map_add(le_to_host16(vhtc->vht_supported_mcs_set.rx_map), WLAN_STA_VHT);
 	blobmsg_close_table(&b, map);
 
-	/* TX map */
 	map = blobmsg_open_table(&b, "tx");
-	hostapd_parse_vht_map_blobmsg(le_to_host16(vhtc->vht_supported_mcs_set.tx_map));
+	mcs_nss_map_add(le_to_host16(vhtc->vht_supported_mcs_set.tx_map), WLAN_STA_VHT);
 	blobmsg_close_table(&b, map);
 
-	blobmsg_close_table(&b, supported_mcs);
+	blobmsg_close_table(&b, mcs_table);
+}
+
+static uint16_t bitfield_value_get(uint8_t *data, uint8_t bit_start, uint8_t bit_count)
+{
+	uint16_t value = 0;
+	uint8_t bit_end = bit_start + bit_count - 1;
+	uint8_t byte_idx = bit_end / 8;
+	uint8_t bit_mask = 1 << (bit_end % 8);
+
+	while (bit_count--) {
+		value = (value << 1) | !!(data[byte_idx] & bit_mask);
+		bit_mask = (bit_mask >> 1) & 0x7f;
+		if (!bit_mask) {
+			bit_mask = 0x80;
+			byte_idx--;
+		}
+	}
+	return value;
 }
 
 static void
-hostapd_parse_capab_blobmsg(struct sta_info *sta)
+capab_table_add(const char *table_name, const bitfield_descriptor *capas,
+		uint8_t capas_len, uint8_t *data, bool filter_false)
 {
-	void *r, *v;
+	void *table = table_name ? blobmsg_open_table(&b, table_name) : NULL;
 
-	v = blobmsg_open_table(&b, "capabilities");
+	for (int i = 0; i < capas_len; i++) {
+		if (capas[i].bitcount > 1) {
+			uint32_t val = bitfield_value_get(data, capas[i].bitstart, capas[i].bitcount);
+			blobmsg_add_u32(&b, capas[i].name, val);
+			continue;
+		}
+
+		uint8_t byte_idx = capas[i].bitstart / 8;
+		uint8_t bit_mask = 1 << (capas[i].bitstart % 8);
+		uint8_t bit_val = !!(data[byte_idx] & bit_mask);
+
+		if (filter_false && !bit_val)
+			continue;
+
+		blobmsg_add_u8(&b, capas[i].name, bit_val);
+	}
+
+	if (table_name)
+		blobmsg_close_table(&b, table);
+}
+
+static void
+ht_mcs_map_add(uint8_t *mcs_bitmask, uint8_t nss, bool ueqm)
+{
+	static const ht_mcs_t ht_mcs[] = {
+		{ "1ss",  0,  7 },
+		{ "2ss",  8, 15 },
+		{ "3ss", 16, 23 },
+		{ "4ss", 24, 31 },
+		{ "1ss",  0,  7 },
+		{ "2ss", 33, 38 },
+		{ "3ss", 39, 52 },
+		{ "4ss", 53, 76 },
+		{ "1ss", 32, 32 }
+	};
+	int mcs_idx = ueqm ? 4 : 0;
+
+	for (int i = 0; i < nss; i++, mcs_idx++) {
+		const ht_mcs_t *mcs = &ht_mcs[mcs_idx];
+		uint8_t byte_idx = mcs->max_mcs / 8;
+		uint8_t bit_mask = 1 << (mcs->max_mcs % 8);
+		int value = -1;
+
+		for (int j = mcs->max_mcs; j >= mcs->min_mcs; j--) {
+			if (mcs_bitmask[byte_idx] & bit_mask) {
+				value = j;
+				break;
+			}
+			bit_mask = (bit_mask >> 1) & 0x7f;
+			if (!bit_mask) {
+				bit_mask = 0x80;
+				byte_idx--;
+			}
+		}
+
+		if (i == 0 && (mcs_bitmask[4] & 0x01))
+			value = HT_MCS_32;
+
+		blobmsg_add_u16(&b, mcs->name, value);
+	}
+}
+
+static void
+ht_capab_add(struct ieee80211_ht_capabilities *htc, bool filter_false)
+{
+	// HT capabilities fields from IEEE 802.11
+	// field values are written in JSON by capab_table_add()
+	// commented out fields are ignored or processed in this function separately
+	static const struct bitfield_descriptor ht_capabilities_info[] = {
+		{ "ldpc_coding", 0, 1 },
+		{ "supported_chan_width_set", 1, 1 },
+		{ "sm_power_save", 2, 2 },
+		{ "ht_greenfield", 4, 1 },
+		{ "short_gi_20mhz", 5, 1 },
+		{ "short_gi_40mhz", 6, 1 },
+		{ "tx_stbc", 7, 1 },
+		{ "rx_stbc", 8, 2 },
+		// { "reserved", 10, 1 },
+		{ "max_a_msdu_len", 11, 1 },
+		{ "dsss_cck_mode_40mhz", 12, 1 },
+		// { "reserved", 13, 1 },
+		{ "forty_mhz_intolerant", 14, 1 }
+		// { "reserved", 15, 1 },
+	};
+	static const struct bitfield_descriptor a_mpdu_params[] = {
+		{ "max_a_mpdu_len_exp", 0, 2 },
+		{ "min_mpdu_start_spacing", 2, 3 }
+		// { "reserved", 5, 3 },
+	};
+	static const struct bitfield_descriptor supported_mcs_set[] = {
+		// { "rx_mcs_bitmask", 0, 77 },
+		// { "reserved", 77, 3 },
+		{ "rx_highest_data_rate", 80, 10 }
+		// { "reserved", 90, 6 },
+		// { "tx_mcs_set_defined", 96, 1 },
+		// { "tx_rx_mcs_set_not_equal", 97, 1 },
+		// { "tx_max_n_spatial_streams", 98, 2 },
+		// { "tx_unequal_modulation", 100, 1 },
+		// { "reserved", 101, 27 },
+	};
+	static const struct bitfield_descriptor ht_ext_capabilities[] = {
+		// { "reserved", 0, 8 },
+		{ "mcs_feedback", 8, 2 },
+		{ "htc-ht", 10, 1 },
+		{ "rd_responder", 11, 1 }
+		// { "reserved", 12, 4 },
+	};
+	static const struct bitfield_descriptor tx_bf_capability_info[] = {
+		{ "implicit_tx_bf_rx", 0, 1 },
+		{ "rx_staggered_sounding", 1, 1 },
+		{ "tx_staggered_sounding", 2, 1 },
+		{ "rx_ndp", 3, 1 },
+		{ "tx_ndp", 4, 1 },
+		{ "implicit_tx_bf", 5, 1 },
+		{ "calibration", 6, 2 },
+		{ "explicit_csi_tx_bf", 8, 1 },
+		{ "explicit_noncompressed_steering", 9, 1 },
+		{ "explicit_compressed_steering", 10, 1 },
+		{ "explicit_tx_bf_csi", 11, 2 },
+		{ "explicit_noncompressed_bf", 13, 2 },
+		{ "explicit_compressed_bf", 15, 2 },
+		{ "min_grouping", 17, 2 },
+		{ "csi_bf_antennas", 19, 2 },
+		{ "noncompressed_steering_bf_antennas", 21, 2 },
+		{ "compressed_steering_bf_antennas", 23, 2 },
+		{ "csi_max_rows_bf", 25, 2 },
+		{ "channel_estimation", 27, 2 }
+		// { "reserved", 29, 3 },
+	};
+	static const struct bitfield_descriptor asel_capabilities[] = {
+		{ "antenna_selection", 0, 1 },
+		{ "explicit_csi_feedback_tx", 1, 1 },
+		{ "antenna_indices_feedback_tx", 2, 1 },
+		{ "explicit_csi_feedback", 3, 1 },
+		{ "antenna_indices_feedback", 4, 1 },
+		{ "receive", 5, 1 },
+		{ "tx_sounding_ppdus", 6, 1 }
+		// { "reserved", 7, 1 },
+	};
+	void *mcs_table, *map;
+	char str[32];
+	uint8_t *mcs_set = htc->supported_mcs_set;
+	bool tx_defined = bitfield_value_get(mcs_set, HT_TX_MCS_SET_DEFINED_BIT, 1) != 0;
+	bool tx_rx_not_equal = tx_defined && bitfield_value_get(mcs_set, HT_TX_RX_MCS_NOT_EQUAL_BIT, 1) != 0;
+	bool tx_ueqm = tx_rx_not_equal && bitfield_value_get(mcs_set, HT_TX_UEQM_BIT, 1) != 0;
+	uint8_t tx_nss = tx_rx_not_equal ? bitfield_value_get(mcs_set, HT_TX_MAX_NSS_BIT, 2) : HT_MAX_SPATIAL_STREAMS - 1;
+
+	capab_table_add("ht_capabilities_info", ht_capabilities_info,
+			ARRAY_SIZE(ht_capabilities_info),
+			(uint8_t *)&htc->ht_capabilities_info, filter_false);
+	capab_table_add("a_mpdu_params", a_mpdu_params,
+			ARRAY_SIZE(a_mpdu_params), &htc->a_mpdu_params, filter_false);
+
+	mcs_table = blobmsg_open_table(&b, "supported_mcs_set");
+
+	wpa_snprintf_hex(str, sizeof(str), mcs_set, IEEE80211_HT_MCS_MASK_LEN);
+	blobmsg_add_string(&b, "rx_mcs_bitmask", str);
+
+	if (tx_defined) {
+		blobmsg_add_u8(&b, "tx_mcs_set_defined", 1);
+		if (tx_rx_not_equal) {
+			blobmsg_add_u8(&b, "tx_rx_mcs_set_not_equal", 1);
+			blobmsg_add_u32(&b, "tx_max_n_spatial_streams", tx_nss + 1);
+			if (tx_ueqm)
+				blobmsg_add_u8(&b, "tx_unequal_modulation", 1);
+		}
+	}
+
+	map = blobmsg_open_table(&b, "rx");
+	ht_mcs_map_add(mcs_set, HT_MAX_SPATIAL_STREAMS, false);
+	blobmsg_close_table(&b, map);
+
+	if (tx_defined) {
+		map = blobmsg_open_table(&b, "tx");
+		ht_mcs_map_add(mcs_set, tx_nss + 1, tx_ueqm);
+		blobmsg_close_table(&b, map);
+	}
+
+	capab_table_add(NULL, supported_mcs_set, ARRAY_SIZE(supported_mcs_set),
+			mcs_set, filter_false);
+	blobmsg_close_table(&b, mcs_table);
+
+	capab_table_add("ht_extended_capabilities", ht_ext_capabilities,
+			ARRAY_SIZE(ht_ext_capabilities),
+			(uint8_t *)&htc->ht_extended_capabilities, filter_false);
+	capab_table_add("tx_bf_capability_info", tx_bf_capability_info,
+			ARRAY_SIZE(tx_bf_capability_info),
+			(uint8_t *)&htc->tx_bf_capability_info, filter_false);
+	capab_table_add("asel_capabilities", asel_capabilities,
+			ARRAY_SIZE(asel_capabilities),
+			&htc->asel_capabilities, filter_false);
+}
+
+static void
+eht_capab_add(struct ieee80211_eht_capabilities *ehtc,
+	      struct ieee80211_he_capabilities *hec, int freq, bool filter_false)
+{
+	void *mcs_table;
+	u8 *ptr;
+	char str[2 + EHT_PPE_THRESH_CAPAB_LEN + 1];
+
+	static const struct bitfield_descriptor mac_capas[] = {
+		{ "epcs_priority_access", 0, 1 },
+		{ "eht_om_control", 1, 1 },
+		{ "txs_mode_1", 2, 1 },
+		{ "txs_mode_2", 3, 1 },
+		{ "restricted_twt", 4, 1 },
+		{ "scs_traffic_descr", 5, 1 },
+		{ "max_mpdu_len", 6, 2 },
+		{ "max_a-mpdu_len_exp", 8, 1 },
+		{ "eht_trs", 9, 1 },
+		{ "txop_return", 10, 1 },
+		{ "two_bqr", 11, 1 },
+		{ "eht_link_adaptation", 12, 2 },
+		{ "unsolicited_epcs_priority",14, 1 }
+		// { "reserved", 15, 1 }
+	};
+
+	static const struct bitfield_descriptor phy_capas[] = {
+		// { "reserved", 0, 1 },
+		{ "320mhz_in_6ghz", 1, 1 },
+		{ "242-tone_ru_in_bw_>20mhz", 2, 1 },
+		{ "ndp_with_4x_eht-ltf", 3, 1 },
+		{ "partial_bw_ul_mu-mimo", 4, 1 },
+		{ "su_beamformer", 5, 1 },
+		{ "su_beamformee", 6, 1 },
+		{ "beamformee_ss_<=80mhz", 7, 3 },
+		{ "beamformee_ss_160mhz", 10, 3 },
+		{ "beamformee_ss_320mhz", 13, 3 },
+		{ "sounding_dimensions_<=80mhz", 16, 3 },
+		{ "sounding_dimensions_160mhz", 19, 3 },
+		{ "sounding_dimensions_320mhz", 22, 3 },
+		{ "ng_=_16_su_feedback", 25, 1 },
+		{ "ng_=_16_mu_feedback", 26, 1 },
+		{ "codebook_size_su_feedback", 27, 1 },
+		{ "codebook_size_mu_feedback", 28, 1 },
+		{ "triggered_su_beamforming_feedback", 29, 1 },
+		{ "triggered_mu_beamforming_feedback", 30, 1 },
+		{ "triggered_cqi_feedback", 31, 1 },
+		{ "partial_bw_dl_mu-mimo", 32, 1 },
+		{ "psr_based_sr_support", 33, 1 },
+		{ "power_boost_factor_ar", 34, 1 },
+		{ "su_mu_ppdu_w_4x_eht_ltf_0_8us_gi", 35, 1 },
+		{ "max_nc", 36, 4 },
+		{ "non_trig_cqi_feedback", 40, 1 },
+		{ "tx_1024_4096_qam<242_tone_ru", 41, 1 },
+		{ "rx_1024_4096qam<242_tone_ru", 42, 1 },
+		{ "ppe_threshold_present", 43, 1 },
+		{ "nominal_packet_padding", 44, 2 },
+		{ "max_eht-ltf", 46, 5 },
+		{ "eht-mcs-15_in_mru", 51, 4 },
+		{ "eth-dup-mcs-14_in_6ghz", 55, 1 },
+		{ "20mhz_sta_recv_ndp_w_wider_bw", 56, 1 },
+		{ "non-ofdma_ul_mu-mimo_<=80mhz", 57, 1 },
+		{ "non-ofdma_ul_mu-mimo_160mhz", 58, 1 },
+		{ "non-ofdma_ul_mu-mimo_320mhz", 59, 1 },
+		{ "mu_beamformer_<=80mhz", 60, 1 },
+		{ "mu_beamformer_160mhz", 61, 1},
+		{ "mu_beamformer_320mhz", 62, 1 },
+		{ "tb_sounding_max_rate", 63, 1 },
+		{ "rx_1024_qam_wider_bw_dl_ofdma", 64, 1 },
+		{ "rx_4096_qam_wider_bw_dl_ofdma", 65, 1 },
+		{ "20mhz-only_limited_capas", 66, 1 },
+		{ "20mhz-only_mu_beamforming", 67, 1 },
+		{ "20mhz-only_mru_support", 68, 1 }
+		// { "reserved", 69, 3 }
+	};
+
+	u8 *he_phy = hec->he_phy_capab_info;
+	u16 ppet_size;
+
+	capab_table_add("mac_capabilities", mac_capas, ARRAY_SIZE(mac_capas),
+			(uint8_t *)&ehtc->mac_cap, filter_false);
+	capab_table_add("phy_capabilities", phy_capas, ARRAY_SIZE(phy_capas),
+			(uint8_t *)&ehtc->phy_cap, filter_false);
+
+	ptr = ehtc->optional;
+	strcpy(str, "0x");
+	mcs_table = blobmsg_open_table(&b, "supported_mcs_set");
+
+	bool band_5g = IS_5GHZ(freq);
+	bool band_6g = freq >= 5895;
+
+	if (band_6g) {
+		wpa_snprintf_hex(str + 2, sizeof(str) - 2, ptr, EHT_PHYCAP_MCS_NSS_LEN_20MHZ_PLUS);
+		blobmsg_add_string(&b, "eht-mcs_map_<80mhz", str);
+		ptr += EHT_PHYCAP_MCS_NSS_LEN_20MHZ_PLUS;
+	} else if (band_5g) {
+		if (bitfield_value_get(he_phy, 2, 3) == 0) {
+			wpa_snprintf_hex(str + 2, sizeof(str) - 2, ptr, EHT_PHYCAP_MCS_NSS_LEN_20MHZ_ONLY);
+			blobmsg_add_string(&b, "eht-mcs_map_20mhz", str);
+			ptr += EHT_PHYCAP_MCS_NSS_LEN_20MHZ_ONLY;
+		} else {
+			wpa_snprintf_hex(str + 2, sizeof(str) - 2, ptr, EHT_PHYCAP_MCS_NSS_LEN_20MHZ_PLUS);
+			blobmsg_add_string(&b, "eht-mcs_map_<80mhz", str);
+			ptr += EHT_PHYCAP_MCS_NSS_LEN_20MHZ_PLUS;
+			if (bitfield_value_get(he_phy, 3, 2)) {
+				wpa_snprintf_hex(str + 2, sizeof(str) - 2, ptr, EHT_PHYCAP_MCS_NSS_LEN_20MHZ_PLUS);
+				blobmsg_add_string(&b, "eht-mcs_map_160mhz", str);
+				ptr += EHT_PHYCAP_MCS_NSS_LEN_20MHZ_PLUS;
+			}
+		}
+	} else {
+		if (bitfield_value_get(he_phy, 1, 1)) {
+			wpa_snprintf_hex(str + 2, sizeof(str) - 2, ptr, EHT_PHYCAP_MCS_NSS_LEN_20MHZ_PLUS);
+			blobmsg_add_string(&b, "eht-mcs_map_<80mhz", str);
+			ptr += EHT_PHYCAP_MCS_NSS_LEN_20MHZ_PLUS;
+		} else {
+			wpa_snprintf_hex(str + 2, sizeof(str) - 2, ptr, EHT_PHYCAP_MCS_NSS_LEN_20MHZ_ONLY);
+			blobmsg_add_string(&b, "eht-mcs_map_20mhz", str);
+			ptr += EHT_PHYCAP_MCS_NSS_LEN_20MHZ_ONLY;
+		}
+	}
+
+	if (band_6g && bitfield_value_get(ehtc->phy_cap, 1, 1)) {
+		wpa_snprintf_hex(str + 2, sizeof(str) - 2, ptr, EHT_PHYCAP_MCS_NSS_LEN_20MHZ_PLUS);
+		blobmsg_add_string(&b, "eht-mcs_map_320mhz", str);
+		ptr += EHT_PHYCAP_MCS_NSS_LEN_20MHZ_PLUS;
+	}
+	blobmsg_close_table(&b, mcs_table);
+
+	ppet_size = smartrg_eht_ppet_size(*(u16 *)ptr, ehtc->phy_cap);
+	if (ppet_size) {
+		wpa_snprintf_hex(str + 2, sizeof(str) - 2, ptr, ppet_size);
+		blobmsg_add_string(&b, "eht_ppe_thresholds", str);
+	}
+}
+
+static void
+he_capab_add(struct ieee80211_he_capabilities *hec, bool filter_false)
+{
+	void *mcs_table, *map;
+
+	static const struct bitfield_descriptor mac_capas[] = {
+		{ "htc_he", 0, 1 },
+		{ "twt_requester", 1, 1 },
+		{ "twt_responder", 2, 1 },
+		// { "dynamic_fragm", 3, 2 },
+		// { "max_n_fragm_msdu", 5, 3 },
+		// { "min_fragment_size", 8, 2 },
+		{ "trigger_frame_mac_padding", 10, 2 },
+		{ "multi_tid_aggregation", 12, 3 },
+		{ "he_link_adaptation", 15, 2 },
+		{ "all_ack", 17, 1 },
+		{ "trs", 18, 1 },
+		{ "bsr", 19, 1 },
+		{ "broadcast_twt", 20, 1 },
+		{ "32_bit_ba_map", 21, 1 },
+		{ "mu_cascading", 22, 1 },
+		{ "ack_enabled_agg", 23,1 },
+		// { "reserved", 24, 1 },
+		{ "om_control", 25, 1 },
+		{ "ofdma_ra", 26, 1 },
+		{ "max_a_mpdu_length_exponent", 27, 2 },
+		{ "a_msdu_frag", 29, 1 },
+		{ "flex_twt_sched", 30, 1 },
+		{ "rx_ctrl_to_mbss", 31, 1 },
+		{ "bsrp_bqrp_a_mpdu_agg", 32, 1 },
+		{ "qtp", 33, 1 },
+		{ "bqr", 34, 1 },
+		{ "psr", 35, 1 },
+		{ "ndp_feedback", 36, 1 },
+		{ "ops", 37, 1 },
+		{ "a_msdu_in_a_mpdu", 38, 1 },
+		{ "multi_tid_aggregation_tx", 39, 3 },
+		{ "he_subchan_select_tx", 42, 1 },
+		{ "ul_2x996_tone_ru", 43, 1 },
+		{ "om_ctrl_ul_mu_disable_rx", 44, 1 },
+		{ "he_dynamic_sm_pwr_save", 45, 1 },
+		{ "punctured_sounding", 46, 1 },
+		{ "ht_vht_trig_frame_rx",47, 1 }
+	};
+
+	static const struct bitfield_descriptor phy_capas[] = {
+		// { "reserved", 0, 1 },
+		{ "channel_width_set", 1, 7 }, // following 7 bits present channel_width_set field
+		{ "40mhz_in_2.4ghz_support", 1, 1 },
+		{ "40_80mhz_in_5ghz_support", 2, 1 },
+		{ "160mhz_in_5ghz_support", 3, 1 },
+		{ "160_80mhz_in_5ghz_support", 4, 1 },
+		{ "242-tones_RU_in_2.4ghz_support", 5, 1 },
+		{ "242-tones_RU_in_5ghz_support", 6, 1 },
+		// { "reserved", 7, 1 },
+		{ "punctured_preamble", 8, 4 },
+		{ "device_class", 12, 1 },
+		{ "ldpc_coding", 13, 1 },
+		{ "he_su_ppdu_w_1x_he_ltf_0_8us_gi", 14, 1 },
+		{ "midamble_rx_max_nsts", 15, 2 },
+		{ "ndp_w_4x_he_ltf_3_2us_gi", 17, 1 },
+		{ "stbc_tx<=80mhz", 18, 1 },
+		{ "stbc_rx<=80mhz", 19, 1 },
+		{ "doppler_tx", 20, 1 },
+		{ "doppler_rx", 21, 1 },
+		{ "full_bw_ul_mu_mimo", 22, 1 },
+		{ "partial_bw_ul_mu_mimo", 23, 1 },
+		{ "dcm_max_constellation_tx", 24, 2 },
+		{ "dcm_max_nss_tx:_1_space_time_stream", 26, 1 },
+		{ "dcm_max_constellation_rx", 27, 2 },
+		{ "dcm_max_nss_rx:_1_space_time_stream", 29, 1 },
+		{ "rx_he_mu_ppdu_fr_non_ap_sta", 30, 1 },
+		{ "su_beamformer", 31, 1 },
+		{ "su_beamformee", 32, 1 },
+		{ "mu_beamformer", 33, 1 },
+		{ "beamformee_sts<=80Mhz", 34, 3 },
+		{ "beamformee_sts>80Mhz", 37, 3 },
+		{ "num_sounding_dimensions<=80mhz", 40, 3 },
+		{ "num_sounding_dimensions>80mhz", 43, 3 },
+		{ "ng_=_16_su_feedback", 46, 1 },
+		{ "ng_=_16_mu_feedback", 47, 1 },
+		{ "codebook_size_su_feedback", 48, 1 },
+		{ "codebook_size_mu_feedback", 49, 1 },
+		{ "triggered_su_beamforming_feedback", 50, 1 },
+		{ "triggered_mu_beamforming_feedback", 51, 1 },
+		{ "triggered_cqi_feedback", 52, 1 },
+		{ "partial_bw_ext_range", 53, 1 },
+		{ "partial_bw_dl_mu_mimo", 54, 1 },
+		{ "ppe_threshold_present", 55, 1 },
+		{ "psr_based_sr_support", 56, 1 },
+		{ "power_boost_factor_ar", 57, 1 },
+		{ "he_su_mu_ppdu_w_4x_he_ltf_0_8us_gi", 58, 1 },
+		{ "max_nc", 59, 3 },
+		{ "stbc_tx>80mhz", 62, 1 },
+		{ "stbc_rx>80mhz", 63, 1 },
+		{ "he_er_su_ppdu_w_4x_he_ltf_0_8us_gi", 64, 1 },
+		{ "20mhz_in_40mhz_he_ppdu_in_2_4ghz", 65, 1 },
+		{ "20mhz_in_160/80_80mhz_he_ppdu", 66, 1 },
+		{ "80mhz_in_160/80_80mhz_he_ppdu", 67, 1 },
+		{ "he_er_su_ppdu_w_1x_he_ltf_0_8us_gi", 68, 1 },
+		{ "midamble_rx_2x_1x_he_ltf", 69, 1 },
+		{ "dcm_max_bw", 70, 2 },
+		{ "longer_than_16_he_sig_b_ofdm_sym", 72, 1 },
+		{ "non_trig_cqi_feedback", 73, 1 },
+		{ "tx_1024_qam<242_tone_ru", 74, 1 },
+		{ "rx_1024_qam<242_tone_ru", 75, 1 },
+		{ "rx_full_bw_su_w_he_mu_ppdu_w_comp_sigb", 76, 1 },
+		{ "rx_full_bw_su_w_he_mu_ppdu_w_non_comp_sigb", 77, 1 },
+		{ "nominal_packet_padding", 78, 2 },
+		{ "he_me_ppdu_w_more_ru_rx_max_n_he_ltf", 80, 1 }
+		// { "reserved", 81, 7 },
+	};
+
+	capab_table_add("mac_capabilities", mac_capas, ARRAY_SIZE(mac_capas),
+			(uint8_t *)&hec->he_mac_capab_info, filter_false);
+	capab_table_add("phy_capabilities", phy_capas, ARRAY_SIZE(phy_capas),
+			(uint8_t *)&hec->he_phy_capab_info, filter_false);
+
+	mcs_table = blobmsg_open_table(&b, "mcs_map");
+
+	map = blobmsg_open_table(&b, "rx");
+	mcs_nss_map_add(le_to_host16(hec->he_basic_supported_mcs_set.rx_map), WLAN_STA_HE);
+	blobmsg_close_table(&b, map);
+
+	map = blobmsg_open_table(&b, "tx");
+	mcs_nss_map_add(le_to_host16(hec->he_basic_supported_mcs_set.tx_map), WLAN_STA_HE);
+	blobmsg_close_table(&b, map);
+
+	blobmsg_close_table(&b, mcs_table);
+}
+
+static void
+sta_capab_add(struct sta_info *sta, int freq, bool filter_false,
+	      struct ieee802_11_elems *elems)
+{
+	void *table, *capab;
+
+	capab = blobmsg_open_table(&b, "capabilities");
 
 	if (sta->vht_capabilities) {
-		r = blobmsg_open_table(&b, "vht");
-		hostapd_parse_vht_capab_blobmsg(sta->vht_capabilities);
-		blobmsg_close_table(&b, r);
+		table = blobmsg_open_table(&b, "vht");
+		vht_capab_add(sta->vht_capabilities, filter_false);
+		blobmsg_close_table(&b, table);
+	} else if (elems && elems->vht_capabilities) {
+		table = blobmsg_open_table(&b, "vht");
+		vht_capab_add((struct ieee80211_vht_capabilities *)elems->vht_capabilities, filter_false);
+		blobmsg_close_table(&b, table);
+	}
+
+	if (sta->he_capab) {
+		table = blobmsg_open_table(&b, "he");
+		he_capab_add(sta->he_capab, filter_false);
+		blobmsg_close_table(&b, table);
+	} else if (elems && elems->he_capabilities &&
+		   elems->he_capabilities_len >= IEEE80211_HE_CAPAB_MIN_LEN &&
+		   elems->he_capabilities_len <= sizeof(struct ieee80211_he_capabilities)) {
+		struct ieee80211_he_capabilities he_tmp;
+
+		os_memset(&he_tmp, 0, sizeof(he_tmp));
+		os_memcpy(&he_tmp, elems->he_capabilities, elems->he_capabilities_len);
+		table = blobmsg_open_table(&b, "he");
+		he_capab_add(&he_tmp, filter_false);
+		blobmsg_close_table(&b, table);
+	}
+
+	if (sta->ht_capabilities) {
+		table = blobmsg_open_table(&b, "ht");
+		ht_capab_add(sta->ht_capabilities, filter_false);
+		blobmsg_close_table(&b, table);
+	} else if (elems && elems->ht_capabilities) {
+		table = blobmsg_open_table(&b, "ht");
+		ht_capab_add((struct ieee80211_ht_capabilities *)elems->ht_capabilities, filter_false);
+		blobmsg_close_table(&b, table);
+	}
+
+	if (sta->eht_capab) {
+		table = blobmsg_open_table(&b, "eht");
+		eht_capab_add(sta->eht_capab, sta->he_capab, freq, filter_false);
+		blobmsg_close_table(&b, table);
+	} else if (elems && elems->eht_capabilities && elems->he_capabilities &&
+		   elems->he_capabilities_len >= IEEE80211_HE_CAPAB_MIN_LEN &&
+		   elems->he_capabilities_len <= sizeof(struct ieee80211_he_capabilities) &&
+		   elems->eht_capabilities_len >= IEEE80211_EHT_CAPAB_MIN_LEN &&
+		   elems->eht_capabilities_len <= sizeof(struct ieee80211_eht_capabilities)) {
+		struct ieee80211_he_capabilities he_tmp;
+		struct ieee80211_eht_capabilities eht_tmp;
+
+		os_memset(&he_tmp, 0, sizeof(he_tmp));
+		os_memcpy(&he_tmp, elems->he_capabilities, elems->he_capabilities_len);
+		os_memset(&eht_tmp, 0, sizeof(eht_tmp));
+		os_memcpy(&eht_tmp, elems->eht_capabilities, elems->eht_capabilities_len);
+		table = blobmsg_open_table(&b, "eht");
+		eht_capab_add(&eht_tmp, &he_tmp, freq, filter_false);
+		blobmsg_close_table(&b, table);
+	}
+
+	blobmsg_close_table(&b, capab);
+}
+
+static void
+sta_vendors_add(struct sta_info *sta)
+{
+	void *table = blobmsg_open_table(&b, "vendors");
+	const struct element *elem;
+	const u8 *ie_buf;
+
+	if (!sta->vendors_ie)
+		goto out;
+
+	ie_buf = wpabuf_head_u8(sta->vendors_ie);
+	for_each_element(elem, ie_buf, sta->vendors_ie->used) {
+		const u8 *oui = elem->data;
+		const u8 *ie_data = elem->data + 3;
+		int data_len = elem->datalen - 3;
+		char oui_str[16];
+		char *out;
+
+		snprintf(oui_str, sizeof(oui_str), "%02x:%02x:%02x",
+			 oui[0], oui[1], oui[2]);
+		out = blobmsg_alloc_string_buffer(&b, oui_str, data_len * 2 + 1);
+		for (int i = 0; i < data_len; i++, out += 2)
+			sprintf(out, "%02x", ie_data[i]);
+		blobmsg_add_string_buffer(&b);
 	}
 
-	/* ToDo: Add HT / HE capability parsing */
+out:
+	blobmsg_close_table(&b, table);
+}
 
-	blobmsg_close_table(&b, v);
+static int compare_char(const void *a, const void *b)
+{
+	return *(char *)a - *(char *)b;
 }
 
 static int
-hostapd_bss_get_clients(struct ubus_context *ctx, struct ubus_object *obj,
+hostapd_bss_get_client_info(struct ubus_context *ctx, struct ubus_object *obj,
 			struct ubus_request_data *req, const char *method,
-			struct blob_attr *msg)
+			struct blob_attr *msg, bool caps_only, bool filter_false)
 {
 	struct hostapd_data *hapd = container_of(obj, struct hostapd_data, ubus.obj);
 	struct hostap_sta_driver_data sta_driver_data;
@@ -255,42 +881,78 @@ hostapd_bss_get_clients(struct ubus_cont
 		{ "authorized", WLAN_STA_AUTHORIZED },
 		{ "preauth", WLAN_STA_PREAUTH },
 		{ "wds", WLAN_STA_WDS },
+		{ "multi_ap", WLAN_STA_MULTI_AP },
 		{ "wmm", WLAN_STA_WMM },
 		{ "ht", WLAN_STA_HT },
 		{ "vht", WLAN_STA_VHT },
 		{ "he", WLAN_STA_HE },
+		{ "eht", WLAN_STA_EHT },
 		{ "wps", WLAN_STA_WPS },
 		{ "mfp", WLAN_STA_MFP },
 	};
 
 	blob_buf_init(&b, 0);
-	blobmsg_add_u32(&b, "freq", hapd->iface->freq);
+	if (!caps_only)
+		blobmsg_add_u32(&b, "freq", hapd->iface->freq);
+
 	list = blobmsg_open_table(&b, "clients");
 	for (sta = hapd->sta_list; sta; sta = sta->next) {
+		struct ieee802_11_elems elems;
+		struct ieee802_11_elems *pelems = NULL;
+		const char *highest_standard;
 		void *r;
 		int i;
 
+		if (sta->assoc_ies &&
+		    ieee802_11_parse_elems(wpabuf_head(sta->assoc_ies),
+					  wpabuf_len(sta->assoc_ies),
+					  &elems, 0) != ParseFailed)
+			pelems = &elems;
+
 		sprintf(mac_buf, MACSTR, MAC2STR(sta->addr));
 		c = blobmsg_open_table(&b, mac_buf);
 		for (i = 0; i < ARRAY_SIZE(sta_flags); i++)
 			blobmsg_add_u8(&b, sta_flags[i].name,
 				       !!(sta->flags & sta_flags[i].flag));
 
+		if (sta->flags & WLAN_STA_EHT)
+			highest_standard = "802.11be";
+		else if (sta->flags & WLAN_STA_HE)
+			highest_standard = "802.11ax";
+		else if (sta->flags & WLAN_STA_VHT)
+			highest_standard = "802.11ac";
+		else if ((sta->flags & WLAN_STA_HT) ||
+			 (pelems && pelems->ht_capabilities))
+			highest_standard = "802.11n";
+		else
+			highest_standard = "802.11b/g";
+		blobmsg_add_string(&b, "highest_standard", highest_standard);
+
 #ifdef CONFIG_MBO
 		blobmsg_add_u8(&b, "mbo", !!(sta->cell_capa));
 #endif
 
+		r = blobmsg_open_array(&b, "supported_rates");
+		if (sta->supported_rates && sta->supported_rates_len > 0) {
+			uint16_t rates_len = sta->supported_rates_len;
+			uint8_t *rate_array = malloc(rates_len);
+			memcpy(rate_array, sta->supported_rates, rates_len);
+			qsort(rate_array, rates_len, sizeof(char), compare_char);
+			for (i = 0; i < rates_len; i++)
+				blobmsg_add_u32(&b, "", (rate_array[i] & 0x7f) >> 1);
+			free(rate_array);
+		}
+		blobmsg_close_array(&b, r);
+
 		r = blobmsg_open_array(&b, "rrm");
 		for (i = 0; i < ARRAY_SIZE(sta->rrm_enabled_capa); i++)
 			blobmsg_add_u32(&b, "", sta->rrm_enabled_capa[i]);
 		blobmsg_close_array(&b, r);
 
 		r = blobmsg_open_array(&b, "extended_capabilities");
-		/* Check if client advertises extended capabilities */
 		if (sta->ext_capability && sta->ext_capability[0] > 0) {
-			for (i = 0; i < sta->ext_capability[0]; i++) {
+			for (i = 0; i < sta->ext_capability[0]; i++)
 				blobmsg_add_u32(&b, "", sta->ext_capability[1 + i]);
-			}
 		}
 		blobmsg_close_array(&b, r);
 
@@ -301,8 +963,8 @@ hostapd_bss_get_clients(struct ubus_cont
 			blobmsg_add_string_buffer(&b);
 #endif
 
-		/* Driver information */
-		if (hostapd_drv_read_sta_data(hapd, &sta_driver_data, sta->addr) >= 0) {
+		if (!caps_only &&
+		    hostapd_drv_read_sta_data(hapd, &sta_driver_data, sta->addr) >= 0) {
 			r = blobmsg_open_table(&b, "bytes");
 			blobmsg_add_u64(&b, "rx", sta_driver_data.rx_bytes);
 			blobmsg_add_u64(&b, "tx", sta_driver_data.tx_bytes);
@@ -316,14 +978,14 @@ hostapd_bss_get_clients(struct ubus_cont
 			blobmsg_add_u32(&b, "tx", sta_driver_data.tx_packets);
 			blobmsg_close_table(&b, r);
 			r = blobmsg_open_table(&b, "rate");
-			/* Rate in kbits */
 			blobmsg_add_u32(&b, "rx", sta_driver_data.current_rx_rate * 100);
 			blobmsg_add_u32(&b, "tx", sta_driver_data.current_tx_rate * 100);
 			blobmsg_close_table(&b, r);
 			blobmsg_add_u32(&b, "signal", sta_driver_data.signal);
 		}
 
-		hostapd_parse_capab_blobmsg(sta);
+		sta_capab_add(sta, hapd->iface->freq, filter_false, pelems);
+		sta_vendors_add(sta);
 
 		blobmsg_close_table(&b, c);
 	}
@@ -334,6 +996,22 @@ hostapd_bss_get_clients(struct ubus_cont
 }
 
 static int
+hostapd_bss_get_clients(struct ubus_context *ctx, struct ubus_object *obj,
+			struct ubus_request_data *req, const char *method,
+			struct blob_attr *msg)
+{
+	return hostapd_bss_get_client_info(ctx, obj, req, method, msg, false, false);
+}
+
+static int
+hostapd_bss_get_client_caps(struct ubus_context *ctx, struct ubus_object *obj,
+			struct ubus_request_data *req, const char *method,
+			struct blob_attr *msg)
+{
+	return hostapd_bss_get_client_info(ctx, obj, req, method, msg, true, true);
+}
+
+static int
 hostapd_bss_get_features(struct ubus_context *ctx, struct ubus_object *obj,
 			struct ubus_request_data *req, const char *method,
 			struct blob_attr *msg)
@@ -1647,6 +2325,7 @@ hostapd_bss_get_sta_ies(struct ubus_cont
 static const struct ubus_method bss_methods[] = {
 	UBUS_METHOD_NOARG("reload", hostapd_bss_reload),
 	UBUS_METHOD_NOARG("get_clients", hostapd_bss_get_clients),
+	UBUS_METHOD_NOARG("get_client_caps", hostapd_bss_get_client_caps),
 #ifdef CONFIG_TAXONOMY
 	UBUS_METHOD("get_sta_ies", hostapd_bss_get_sta_ies, addr_policy),
 #endif
--- a/src/ap/ieee802_11_eht.c
+++ b/src/ap/ieee802_11_eht.c
@@ -46,6 +46,9 @@ static u16 ieee80211_eht_ppet_size(u16 p
 	return sz;
 }
 
+u16 smartrg_eht_ppet_size(u16 ppe_thres_hdr, const u8 *phy_cap_info) {
+	return ieee80211_eht_ppet_size(ppe_thres_hdr, phy_cap_info);
+}
 
 static u8 ieee80211_eht_mcs_set_size(enum hostapd_hw_mode mode, u8 opclass,
 				     const u8 *he_phy_cap,
